[Fleet] Redesign agent flyout

criamico opened the pull request:
## Summary
Redesign of agent enrollment flyout: https://github.com/elastic/kibana/issues/125534

- Try to enroll a new agent, the new UI should be like this:

![Screenshot 2022-03-24 at 18 32 04](https://user-images.githubusercontent.com/16084106/159976957-b426fda1-6cc4-46db-8b31-7074bccd74e4.png)

![Screenshot 2022-03-24 at 18 39 26](https://user-images.githubusercontent.com/16084106/159977244-d5e60f94-c46e-451f-9836-67a6617d885b.png)
![Screenshot 2022-03-24 at 18 33 33](https://user-images.githubusercontent.com/16084106/159977351-929a2d8b-5b26-48fa-a96e-aafc85c5af29.png)
![Screenshot 2022-04-06 at 17 21 28](https://user-images.githubusercontent.com/16084106/162015788-7f359ac6-731d-4fe4-bdde-246e5f65ebbb.png)


### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:



### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

joshdover commented:
Submitted a PR to fix a few things here: https://github.com/criamico/kibana/pull/1

criamico commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

juliaElastic commented on x-pack/plugins/fleet/public/components/agent_enrollment_flyout/instructions.tsx:
```diff
@@ -0,0 +1,374 @@
+  return <FleetServerRequirementPage />;
+};
+
+export const Instructions = React.memo<InstructionProps>(
```
I think the name `Instructions` is a little too generic. What about `AgentEnrollmentInstructions`?

juliaElastic commented:
I noticed a broken feature in Add agent flyout, with Fleet Server instructions:
- Create a Fleet Server policy
- Bug: the created policy doesn't appear in the dropdown
- Expected: the created Fleet Server policy should appear selected in agent policy dropdown
<img width="885" alt="image" src="https://user-images.githubusercontent.com/90178898/160366675-548156c3-0aea-4b53-9615-9aaf45931d60.png">

juliaElastic commented:
After enrolling a Fleet Server, and opening Add agent flyout, I see failing agent_status queries with this error:

<img width="1286" alt="image" src="https://user-images.githubusercontent.com/90178898/160372743-4004438b-3966-4fe5-8362-acd5558f849d.png">
```
Request URL: http://localhost:5620/api/fleet/agent_status/data
Request Method: GET
Status Code: 400 Bad Request
{"statusCode":400,"error":"Bad Request","message":"[request query.agentsIds]: expected at least one defined value but got [undefined]"}
```

criamico commented:
@juliaElastic Thanks for testing it out! I'm looking at these issues.

juliaElastic commented on x-pack/plugins/fleet/public/components/agent_enrollment_flyout/agent_policy_select_create.tsx:
```diff
@@ -16,43 +16,60 @@ import { AgentPolicyCreateInlineForm } from '../../applications/fleet/sections/a
   agentPolicies,
   excludeFleetServer,
-  onAgentPolicyChange,
+  setSelectedPolicyId,
```
`setSelectedPolicyId` is not always set when using `<SelectCreateAgentPolicy>` , maybe that is causing the undefined error

criamico replied:
I think you are right there, I think I fixed it

criamico commented:
@juliaElastic would you mind testing again those issue? I pushed a commit that should fix them.

criamico commented:
@elasticmachine merge upstream

juliaElastic commented:
> @juliaElastic would you mind testing again those issues? I pushed a commit that should fix them.

Thanks, the fleet server enroll flow is fixed now.
I'm still seeing the error with the `/agent_status/data` call though.

juliaElastic commented:
I tested the flow of enrolling agent to Fleet, and I see the UI updated to show that 1 agent has enrolled, and incoming data is confirmed. A few observations:
- `Confirm agent Enrollment` step spinner keeps spinning even after agents has been enrolled. Is this expected?
- The queries to check for agents/agent data keep going on in the background. Is it expected or should it stop at some point?
- The `Confirm incoming data` step is called `Incoming data confirmed` in the original issue, which is the final one? https://github.com/elastic/kibana/issues/125534

<img width="1140" alt="image" src="https://user-images.githubusercontent.com/90178898/160388116-afab4476-9b71-45c8-a12c-e51f0c73ad57.png">

juliaElastic commented on x-pack/plugins/fleet/public/components/agent_enrollment_flyout/confirm_agent_enrollment.tsx:
```diff
@@ -5,39 +5,97 @@
+    return (
+      <EuiText>
+        <FormattedMessage
+          id="xpack.fleet.enrollmentInstructions.troubleshootingText"
```
nit: the troubleshooting link could be extracted to a small component

juliaElastic commented on x-pack/plugins/fleet/public/components/enrollment_instructions/manual/index.tsx:
```diff
@@ -51,15 +54,43 @@ sudo elastic-agent enroll ${enrollArgs} \nsudo systemctl enable elastic-agent \n
 sudo elastic-agent enroll ${enrollArgs} \nsudo systemctl enable elastic-agent \nsudo systemctl start elastic-agent`;
 
+  const link1 = '';
+  const link2 = '';
```
nit: `link1` and `link2` could be renamed to more meaningful variable names

juliaElastic commented on x-pack/plugins/fleet/public/components/agent_enrollment_flyout/confirm_agent_enrollment.tsx:
```diff
@@ -5,39 +5,97 @@
+ * @param policyId
+ * @returns
+ */
+export const usePollingAgentCount = (policyId: string | undefined) => {
```
unit tests are missing, I think you can add them in another pr

criamico commented:
> I tested the flow of enrolling agent to Fleet, and I see the UI updated to show that 1 agent has enrolled, and incoming data is confirmed. A few observations:
> 
> * `Confirm agent Enrollment` step spinner keeps spinning even after agents has been enrolled. Is this expected?

No, I still have to fix the state of the last two steps and the spinners. I'll check it.

> * The queries to check for agents/agent data keep going on in the background, even after closing the flyout. Is it expected or should it stop at some point?

No it should stop when the flyout closes. It is a bug.

> * The `Confirm incoming data` step is called `Incoming data confirmed` in the original issue, which is the final one? [[Fleet] Agent Install Flyout Redesign #125534](https://github.com/elastic/kibana/issues/125534)

 I think it's supposed to be `Confirm incoming data` when loading and `Incoming data confirmed`. I'll fix it along with the first point.

![Screenshot 2022-03-28 at 14 46 41](https://user-images.githubusercontent.com/16084106/160401598-07462f8f-69c6-4d45-a000-8cad3551ae66.png)

criamico commented:
I pushed a commit where I addressed most of the problems, however I still need to fix the endpoints that keep getting called after closing the panel.

nchaulet commented on x-pack/plugins/fleet/public/components/agent_enrollment_flyout/confirm_agent_enrollment.tsx:
```diff
@@ -5,39 +5,108 @@
+  const mountedAt = useRef(Date.now());
+
+  useEffect(() => {
+    let timeout: NodeJS.Timeout;
+
+    const poll = () => {
+      timeout = setTimeout(async () => {
+        const secSinceMounted = Math.ceil((Date.now() - mountedAt.current) / 1000);
+        const request = await sendGetAgents({
+          kuery: `${AGENTS_PREFIX}.policy_id:"${policyId}" and ${AGENTS_PREFIX}.enrolled_at >= "now-${secSinceMounted}s"`,
+          showInactive: false,
+        });
+        const newAgentIds = request.data?.items.map((i) => i.id) ?? agentIds;
+
+        if (newAgentIds.some((id) => !agentIds.includes(id))) {
+          setAgentIds(newAgentIds);
+        }
+
+        poll();
+      }, POLLING_INTERVAL_MS);
+    };
+
+    poll();
+    return () => {
+      clearTimeout(timeout);
+    };
```
There is a potential race condition here that could lead to an infinite polling, I think it could be solved by introducing an extra variable.

```suggestion
    let timeout: NodeJS.Timeout;
    let isAborted: boolean;

    const poll = () => {
      timeout = setTimeout(async () => {
        const secSinceMounted = Math.ceil((Date.now() - mountedAt.current) / 1000);
        const request = await sendGetAgents({
          kuery: `${AGENTS_PREFIX}.policy_id:"${policyId}" and ${AGENTS_PREFIX}.enrolled_at >= "now-${secSinceMounted}s"`,
          showInactive: false,
        });
        const newAgentIds = request.data?.items.map((i) => i.id) ?? agentIds;

        if (newAgentIds.some((id) => !agentIds.includes(id))) {
          setAgentIds(newAgentIds);
        }
        if (!isAborted) {
          poll();
        }
      }, POLLING_INTERVAL_MS);
    };

    poll();
    return () => {
      isAborted = true;
      clearTimeout(timeout);
    };
```

kpollich reviewed (commented):
It seems like after creating one agent, navigating back into the flyout incurs this "empty state" in the dropdown. I think this is similar to the issue @juliaElastic reported above.

![image](https://user-images.githubusercontent.com/6766512/160466930-1dd84105-ef37-4e90-aa83-84565805bf14.png)

juliaElastic commented:
I noticed that when there is no Fleet Server enrolled initially, there is no way to see the Agent standalone instructions (which doesn't require Fleet Server).
Is this going to be solved by the [second part](https://github.com/elastic/kibana/issues/125640)? If so, should the 2 redesigns be included in the same release?
<img width="522" alt="image" src="https://user-images.githubusercontent.com/90178898/160558933-7f5e2e6a-5e68-4fb3-8c33-92e42a6f9aee.png">

juliaElastic commented:
> It seems like after creating one agent, navigating back into the flyout incurs this "empty state" in the dropdown. I think this is similar to the issue @juliaElastic reported above.
> 
> ![image](https://user-images.githubusercontent.com/6766512/160466930-1dd84105-ef37-4e90-aa83-84565805bf14.png)

@kpollich I couldn't reproduce this, do you have only one policy in the dropdown? It is expected to have an empty dropdown if you have at least 2 policies (e.g. fleet server policy, agent policy)

criamico commented:
> I noticed that when there is no Fleet Server enrolled initially, there is no way to see the Agent standalone instructions (which doesn't require Fleet Server). Is this going to be solved by the [second part]

That's a good question, originally the idea was to release everything on 8.2, but given that FF is today I'm not sure if @kpollich will have time to rebase his PR on mine and release his part. 
I'll try to fix as many bugs as possible so if there's enough time for that part it can be released on its own.

joshdover commented:
Good catch @juliaElastic. I agree we shouldn't merge this if it is going to break the flow. Whether or not FF is today, whatever is on `main` should always work else it will impact testers and other teams within the organization.

So we have two options:
1. Attempt to rebase https://github.com/elastic/kibana/pull/127786 and merge them to `main` as a single PR
2. Restore the previous standalone agent functionality when there's no Fleet Server configured in this PR and merge this.

IMO option (2) is probably the only option that results in us shipping any of these changes in 8.2. @criamico do you have an idea of the amount of effort required to make this happen?

kpollich commented:
> @kpollich I couldn't reproduce this, do you have only one policy in the dropdown? It is expected to have an empty dropdown if you have at least 2 policies (e.g. fleet server policy, agent policy)

@juliaElastic Here are my steps to replicate the issue:

1. Set up a Fleet Server
2. Open the "Add Agent" flyout and select the "Create policy" option next to the agent policy dropdown
3. Once a policy is created, close the flyout
4. Re-open the flyout and note the state of the dropdown

https://user-images.githubusercontent.com/6766512/160608572-c5e3f6f8-ef8f-4627-8772-feb688ef7922.mov

It seems that there's an option selected when I click into the dropdown, but the option isn't appearing as the value in the `EuiSelect` component.

kpollich commented:
> ...given that FF is today I'm not sure if @kpollich will have time to rebase his PR on mine and release his part.
I'll try to fix as many bugs as possible so if there's enough time for that part it can be released on its own.

I think it's a safe assumption that https://github.com/elastic/kibana/pull/127786 will not be able to land before FF. It's a large PR that needs to be integrated with @criamico's set of changes here and then reviewed after that process. I think the risk of combining these PRs and reviewing/merging the rather large combined changeset before FF today is a bit too high. 

@joshdover's option number two above does seem to be our best chance of landing some of this behavior in 8.2. 

> Restore the previous standalone agent functionality when there's no Fleet Server configured in this PR and merge this.

I believe we need to restore the `standalone_instructions.tsx` [file](https://github.com/elastic/kibana/pull/128381/files#diff-a5a9e6507520a4c80fef336a388acbe876600c54eaf9279114e115f11da9270e) in this PR to facilitate this. Previously we were rendering this component [here](https://github.com/elastic/kibana/pull/128381/files#diff-8cadc4af2e577cf61f0383ef0ac55f45e37fe1f67ac1a2b17129f1be1a58443aL165-R146).

juliaElastic commented:
> > @kpollich I couldn't reproduce this, do you have only one policy in the dropdown? It is expected to have an empty dropdown if you have at least 2 policies (e.g. fleet server policy, agent policy)
> 
> @juliaElastic Here are my steps to replicate the issue:
> 
> 1. Set up a Fleet Server
> 2. Open the "Add Agent" flyout and select the "Create policy" option next to the agent policy dropdown
> 3. Once a policy is created, close the flyout
> 4. Re-open the flyout and note the state of the dropdown
> 
> It seems that there's an option selected when I click into the dropdown, but the option isn't appearing as the value in the `EuiSelect` component.

@kpollich This is not an issue, I think. Since [this](https://github.com/elastic/kibana/pull/121628) change it is expected that a policy is not selected by default if more than one policies exist.

criamico commented:
@elasticmachine merge upstream

criamico commented:
@elasticmachine merge upstream

criamico commented:
Opening up the PR for review again for 8.2. Most of the previous issues have been fixed and the functionality around fleet server flow has been restored.

kpollich commented:
Posting my manual testing results running locally/on-prem here

## Standalone

**Questions**:
- Should the "Configure the agent step" have an initial "loading" state upon opening the flyout? 

**Works as expected**:
- [x] Policy creation
  - Minor caveat that @criamico and I discussed offline - the "success" callout is missing here. Let's make sure we file a follow up issue to capture replacing it.
- [x]  Policy object appears in YML block
- [x] New agent enrollment command is properly generated

## Managed

**Questions**:
- Should the "Continue" button after setting up Fleet Server open the "Add Agent" flyout instead of closing the flyout and forcing the user to reopen it via the "Add Agent" button?

**Works as expected**:
- [x] Prompted to set up Fleet Server when one is not detected
- [x] Fleet Server policy creation
- [x] Fleet Server installation command appears in code block
- [x] Once Fleet Server is online, prompted to continue
- [x] Prompted to select policy from dropdown
- [x] New radio selection UI for Managed/Standalone enrollment appears
- [x] New agent installation command is present
- [x] Once agent is running, enrollment is confirmed in UI

I can't seem to get the "Confirm incoming data working" with my dockerized Agent/Fleet Server setup. I'll reach out offline to maybe take a look at that further.

kpollich commented:
@elasticmachine merge upstream

kpollich commented:
I added the `ci:deploy-cloud` label to try and test this on cloud but it doesn't seem like it took. Merging upstream and trying again.

juliaElastic commented on x-pack/plugins/fleet/public/components/enrollment_instructions/standalone/index.tsx:
```diff
@@ -0,0 +1,48 @@
+
+  const STANDALONE_RUN_INSTRUCTIONS_WINDOWS = `wget https://artifacts.elastic.co/downloads/beats/elastic-agent/elastic-agent-${kibanaVersion}-windows-x86_64.zip -OutFile elastic-agent-${kibanaVersion}-windows-x86_64.zip
+  Expand-Archive .\elastic-agent-${kibanaVersion}-windows-x86_64.zip
+  .\\elastic-agent.exe install`;
```
Can you add the missing `cd` command that was fixed for manual instructions here: https://github.com/elastic/kibana/pull/129051/files

criamico replied:
thanks, I missed this!

juliaElastic commented:
I tested today, and I found that the "Confirm incoming data" step is not changing to successful, even though I see Agent logs after some time in Fleet UI. On the Network tab I don't see the `data?agentIds` call repeated, maybe that is the reason the flyout doesn't see the incoming data?

This is an on prem setup with Fleet Server running with docker and Elastic Agent installed locally.

<img width="1793" alt="image" src="https://user-images.githubusercontent.com/90178898/162186807-4c2882d3-cd59-462d-91fd-46c460ad6deb.png">

juliaElastic commented:
> I noticed one more broken feature:
> 
> * enroll fleet server and then unenroll/stop it
> * there is a callout that appears to show Fleet Server instructions again
> * when clicking on the Add Fleet Server button, the flyout shows add agent instructions
> * expected: to see Fleet Server instructions
> 
> <img alt="image" width="1489" src="https://user-images.githubusercontent.com/90178898/160623866-3652b8ee-d14a-4231-b182-5aa12e4f1b49.png">

Tested this again and it is solved now.

juliaElastic commented:
> Did you check that the agent policy that you selected on the flyout dropdown is the same as the agent that you are running? I think that otherwise it doesn't show up, since it queries for that agent policy id.
> 
> @kpollich Could you also try it out? I'm trying to understand if the problem here is the selected policy or the logic in `usePollingAgentCount` that only checks the agents enrolled since the panel was open. This would in theory guarantee that we only show the agent the first time that we open it, but it might also cause some bugs.

Yes, I am enrolling an agent to the policy which the flyout has selected.

I think the problem is that the code is only querying the incoming data once, and doesn't keep querying it, so doesn't notice when the data appears: https://github.com/elastic/kibana/pull/128381/files#diff-bc5b344af51d4fba09dfaf95b51505b18d2b33e0863d0d1cee125429edb87341

If I make the query manually after some time, it shows that there is data:
```
http://localhost:5620/api/fleet/agent_status/data?agentsIds=ef380e3b-7422-427a-8aa1-852cf97b8211
{
items: [
    {
    ef380e3b-7422-427a-8aa1-852cf97b8211: {
              data: true
            }
        }
    ]
}
```

criamico commented:
I added another commit where I introduced a retry logic for incoming data step. Now the `agent_status/data` is fetched every 5 seconds until the data is retrieved.

criamico commented:
@elasticmachine merge upstream

juliaElastic commented:
> I added another commit where I introduced a retry logic for incoming data step. Now the `agent_status/data` is fetched every 5 seconds until the data is retrieved.

I've tried to verify this, however I still don't see the data calls being repeated.

<img width="1772" alt="image" src="https://user-images.githubusercontent.com/90178898/162700191-17cf5d16-db71-4ee5-8527-d413b70e999e.png">

juliaElastic commented on x-pack/plugins/fleet/public/components/agent_enrollment_flyout/use_get_agent_incoming_data.tsx:
```diff
@@ -49,30 +33,81 @@ export const useGetAgentIncomingData = (
+    if (isAborted || incomingData.length > 0) clearTimeout(timeout.current);
+
+    return () => {
+      isAborted = true;
```
I'm not sure when this callback function is executed, but in my local I see `isAborted` set to `true` right after the first invocation of `poll()`.

juliaElastic replied:
this might be the problematic logic:

```
setIncomingData(data?.items);
...
if (incomingData.length > 0) clearTimeout();
```

In my local, the `data?agentsIds` request comes back with the agent id inside `items` with `data: false` at first, I think the timeout shouldn't be cleared in that case.

```
{
items: [
{
cb674401-fd99-4c19-8e24-4fe4604ec4b6: {
data: false
}
}
]
}
```

juliaElastic commented:
I have an observation about the overall "Confirm enrolled agent" and "Confirm incoming data" feature. 

Sometimes I closed the flyout accidentally or to check another view of Agents, and there is no way to go back to the state of the flyout which confirms the data, since the queries only check for agents added since opening the flyout.
This means it is easy to lose the information on the latest agent, which might be annoying to users.

This is probably out of scope for this issue, but I think it would be nice to improve the situation to check the confirm state for any recently added agents, even after closing the flyout.

criamico commented:
> I have an observation about the overall "Confirm enrolled agent" and "Confirm incoming data" feature.
> 
> Sometimes I closed the flyout accidentally or to check another view of Agents, and there is no way to go back to the state of the flyout which confirms the data, since the queries only check for agents added since opening the flyout. This means it is easy to lose the information on the latest agent, which might be annoying to users.
> 
> This is probably out of scope for this issue, but I think it would be nice to improve the situation to check the confirm state for any recently added agents, even after closing the flyout.

I know, I also feel that it is confusing for the user since it could prevent the new agents from showing up. This functionality was added after a discussion with @joshdover. 
Since it is quite easy to remove and it wasn't explicitly requested by a PM, I would remove it from this PR and later we might reevaluate it.

joshdover commented:
> I know, I also feel that it is confusing for the user since it could prevent the new agents from showing up. This functionality was added after a discussion with @joshdover.
> Since it is quite easy to remove and it wasn't explicitly requested by a PM, I would remove it from this PR and later we might reevaluate it.

IMO it's more confusing to open the flyout and it already have this bright green success UI shown before the user has done anything?

<img width="438" alt="image" src="https://user-images.githubusercontent.com/1813008/162726112-d6561968-416f-4aad-aaf5-110449927b1f.png">

Maybe we could enhance this in the future to have 2 separate indicators:
- One big and obvious success confirmation similar to the one in this PR
- A smaller indicator / link like "See X other agents enrolled in the last 15 minutes ->"

joshdover commented:
BTW feel free to merge without my approval, I may not get to this today.

juliaElastic commented:
> IMO it's more confusing to open the flyout and it already have this bright green success UI shown before the user has done anything?

The confirmation will only show up if someone enrolled an agent to the selected agent policy in the last few minutes. That is unlikely to be done by someone else, isn't it?

juliaElastic commented:
should the labels be updated to `v8.3`?

criamico commented:
> IMO it's more confusing to open the flyout and it already have this bright green success UI shown before the user has done anything?
> 
> <img alt="image" width="438" src="https://user-images.githubusercontent.com/1813008/162726112-d6561968-416f-4aad-aaf5-110449927b1f.png">
> 
> Maybe we could enhance this in the future to have 2 separate indicators:
> 
> * One big and obvious success confirmation similar to the one in this PR
> * A smaller indicator / link like "See X other agents enrolled in the last 15 minutes ->"

I think that this piece of functionality needs further investigation and possibly involvement of a PM. It is a good candidate for an enhancement ticket.

kpollich commented:
I'm having trouble getting passed the "Confirm incoming data" step in Cloud. 

I'm enrolling an agent in a basic Agent Policy that just includes the `system` integration, and I'm running the agent on my Macbook via docker, e.g.

```
docker run --rm --add-host elasticsearch.local:192.168.65.2 \
  -e FLEET_URL=https://6e3f3c836ed0d22d3db21117a0894f2f.fleet.us-west2.gcp.elastic-cloud.com:443\
  -e FLEET_ENROLL=true \
  -e FLEET_ENROLLMENT_TOKEN=MThjV0JJQUJFTENteXpEd0xBTDI6ZUZSRHpvVmlTUGVPWVFHWkhHTFZ3Zw== \
  docker.elastic.co/beats/elastic-agent:8.2.0-SNAPSHOT
```

![image](https://user-images.githubusercontent.com/6766512/162747506-7752677e-ed10-4605-aecf-05e4c9b84e3e.png)

My agent starts up as expected, and Fleet reports that it's enrolled, but the "Confirm incoming data" spinner never goes away. I let it sit for 15ish minutes with no change. 

Here's what I think is the relevant polling request from my network tab:

<details>
<summary><code>/api/fleet/agents?page=1&perPage=20&showInactive=false&showUpgradeable=false</code></summary>

```json
{
    "list": [
        {
            "id": "a7c7bb99-504f-4d4b-bd09-10c2fce9ae1b",
            "access_api_key_id": "t3bDGIABELCmyzDwjYIP",
            "action_seq_no": [
                -1
            ],
            "active": true,
            "agent": {
                "id": "a7c7bb99-504f-4d4b-bd09-10c2fce9ae1b",
                "version": "8.2.0"
            },
            "enrolled_at": "2022-04-11T13:15:10Z",
            "local_metadata": {
                "elastic": {
                    "agent": {
                        "build.original": "8.2.0-SNAPSHOT (build: 4b4938b6413972b890ea2f8d256b0e066deea109 at 2022-04-07 15:12:02 +0000 UTC)",
                        "id": "a7c7bb99-504f-4d4b-bd09-10c2fce9ae1b",
                        "log_level": "info",
                        "snapshot": true,
                        "upgradeable": false,
                        "version": "8.2.0"
                    }
                },
                "host": {
                    "architecture": "x86_64",
                    "hostname": "6a2804ec7569",
                    "id": "",
                    "ip": [
                        "127.0.0.1/8",
                        "172.17.0.3/16"
                    ],
                    "mac": [
                        "02:42:ac:11:00:03"
                    ],
                    "name": "6a2804ec7569"
                },
                "os": {
                    "family": "debian",
                    "full": "Ubuntu focal(20.04.4 LTS (Focal Fossa))",
                    "kernel": "5.10.76-linuxkit",
                    "name": "Ubuntu",
                    "platform": "ubuntu",
                    "version": "20.04.4 LTS (Focal Fossa)"
                }
            },
            "policy_id": "b524cbd0-b671-11ec-9278-35137f95496f",
            "type": "PERMANENT",
            "updated_at": "2022-04-11T13:19:22Z",
            "last_checkin_status": "online",
            "last_checkin": "2022-04-11T13:19:18Z",
            "default_api_key": "uXbDGIABELCmyzDwloJS:0pk0edMnSTeiCyi6CxG14A",
            "default_api_key_id": "uXbDGIABELCmyzDwloJS",
            "policy_output_permissions_hash": "a98b2704d96f3a72c77ee2e499898bc091d3c710d56a7b75c9f7f7b7b0f14f7b",
            "policy_revision_idx": 1,
            "policy_coordinator_idx": 1,
            "policy_revision": 1,
            "status": "online",
            "packages": []
        },
        {
            "id": "d56c6900-ac4d-4798-829d-e60d12753cc7",
            "access_api_key_id": "J0tFAYABicK5TYBjdnJC",
            "action_seq_no": [
                -1
            ],
            "active": true,
            "agent": {
                "id": "d56c6900-ac4d-4798-829d-e60d12753cc7",
                "version": "8.3.0"
            },
            "enrolled_at": "2022-04-06T23:46:11Z",
            "local_metadata": {
                "elastic": {
                    "agent": {
                        "build.original": "8.3.0-SNAPSHOT (build: 12eb476712eed1b9b6b94f2b8c807df4ccac3b3b at 2022-04-06 16:14:16 +0000 UTC)",
                        "id": "d56c6900-ac4d-4798-829d-e60d12753cc7",
                        "log_level": "info",
                        "snapshot": true,
                        "upgradeable": false,
                        "version": "8.3.0"
                    }
                },
                "host": {
                    "architecture": "x86_64",
                    "hostname": "372f76872ca3",
                    "id": "",
                    "ip": [
                        "127.0.0.1/8",
                        "172.17.0.99/16"
                    ],
                    "mac": [
                        "02:42:ac:11:00:63"
                    ],
                    "name": "372f76872ca3"
                },
                "os": {
                    "family": "debian",
                    "full": "Ubuntu focal(20.04.4 LTS (Focal Fossa))",
                    "kernel": "4.15.0-1032-gcp",
                    "name": "Ubuntu",
                    "platform": "ubuntu",
                    "version": "20.04.4 LTS (Focal Fossa)"
                }
            },
            "policy_id": "policy-elastic-agent-on-cloud",
            "type": "PERMANENT",
            "default_api_key_id": "FrZMAYABELCmyzDwcwpl",
            "policy_output_permissions_hash": "1c50ec8787bd4cae8cb3e1325361e3ba281f88b698f9c401c0fd7846ff2aefb3",
            "default_api_key": "FrZMAYABELCmyzDwcwpl:_I6_CL8tR7aQ6QJEZcMxhg",
            "policy_revision_idx": 4,
            "policy_coordinator_idx": 1,
            "updated_at": "2022-04-11T13:19:42Z",
            "last_checkin_status": "online",
            "last_checkin": "2022-04-11T13:19:35Z",
            "policy_revision": 4,
            "status": "online",
            "packages": []
        }
    ],
    "items": [
        {
            "id": "a7c7bb99-504f-4d4b-bd09-10c2fce9ae1b",
            "access_api_key_id": "t3bDGIABELCmyzDwjYIP",
            "action_seq_no": [
                -1
            ],
            "active": true,
            "agent": {
                "id": "a7c7bb99-504f-4d4b-bd09-10c2fce9ae1b",
                "version": "8.2.0"
            },
            "enrolled_at": "2022-04-11T13:15:10Z",
            "local_metadata": {
                "elastic": {
                    "agent": {
                        "build.original": "8.2.0-SNAPSHOT (build: 4b4938b6413972b890ea2f8d256b0e066deea109 at 2022-04-07 15:12:02 +0000 UTC)",
                        "id": "a7c7bb99-504f-4d4b-bd09-10c2fce9ae1b",
                        "log_level": "info",
                        "snapshot": true,
                        "upgradeable": false,
                        "version": "8.2.0"
                    }
                },
                "host": {
                    "architecture": "x86_64",
                    "hostname": "6a2804ec7569",
                    "id": "",
                    "ip": [
                        "127.0.0.1/8",
                        "172.17.0.3/16"
                    ],
                    "mac": [
                        "02:42:ac:11:00:03"
                    ],
                    "name": "6a2804ec7569"
                },
                "os": {
                    "family": "debian",
                    "full": "Ubuntu focal(20.04.4 LTS (Focal Fossa))",
                    "kernel": "5.10.76-linuxkit",
                    "name": "Ubuntu",
                    "platform": "ubuntu",
                    "version": "20.04.4 LTS (Focal Fossa)"
                }
            },
            "policy_id": "b524cbd0-b671-11ec-9278-35137f95496f",
            "type": "PERMANENT",
            "updated_at": "2022-04-11T13:19:22Z",
            "last_checkin_status": "online",
            "last_checkin": "2022-04-11T13:19:18Z",
            "default_api_key": "uXbDGIABELCmyzDwloJS:0pk0edMnSTeiCyi6CxG14A",
            "default_api_key_id": "uXbDGIABELCmyzDwloJS",
            "policy_output_permissions_hash": "a98b2704d96f3a72c77ee2e499898bc091d3c710d56a7b75c9f7f7b7b0f14f7b",
            "policy_revision_idx": 1,
            "policy_coordinator_idx": 1,
            "policy_revision": 1,
            "status": "online",
            "packages": []
        },
        {
            "id": "d56c6900-ac4d-4798-829d-e60d12753cc7",
            "access_api_key_id": "J0tFAYABicK5TYBjdnJC",
            "action_seq_no": [
                -1
            ],
            "active": true,
            "agent": {
                "id": "d56c6900-ac4d-4798-829d-e60d12753cc7",
                "version": "8.3.0"
            },
            "enrolled_at": "2022-04-06T23:46:11Z",
            "local_metadata": {
                "elastic": {
                    "agent": {
                        "build.original": "8.3.0-SNAPSHOT (build: 12eb476712eed1b9b6b94f2b8c807df4ccac3b3b at 2022-04-06 16:14:16 +0000 UTC)",
                        "id": "d56c6900-ac4d-4798-829d-e60d12753cc7",
                        "log_level": "info",
                        "snapshot": true,
                        "upgradeable": false,
                        "version": "8.3.0"
                    }
                },
                "host": {
                    "architecture": "x86_64",
                    "hostname": "372f76872ca3",
                    "id": "",
                    "ip": [
                        "127.0.0.1/8",
                        "172.17.0.99/16"
                    ],
                    "mac": [
                        "02:42:ac:11:00:63"
                    ],
                    "name": "372f76872ca3"
                },
                "os": {
                    "family": "debian",
                    "full": "Ubuntu focal(20.04.4 LTS (Focal Fossa))",
                    "kernel": "4.15.0-1032-gcp",
                    "name": "Ubuntu",
                    "platform": "ubuntu",
                    "version": "20.04.4 LTS (Focal Fossa)"
                }
            },
            "policy_id": "policy-elastic-agent-on-cloud",
            "type": "PERMANENT",
            "default_api_key_id": "FrZMAYABELCmyzDwcwpl",
            "policy_output_permissions_hash": "1c50ec8787bd4cae8cb3e1325361e3ba281f88b698f9c401c0fd7846ff2aefb3",
            "default_api_key": "FrZMAYABELCmyzDwcwpl:_I6_CL8tR7aQ6QJEZcMxhg",
            "policy_revision_idx": 4,
            "policy_coordinator_idx": 1,
            "updated_at": "2022-04-11T13:19:42Z",
            "last_checkin_status": "online",
            "last_checkin": "2022-04-11T13:19:35Z",
            "policy_revision": 4,
            "status": "online",
            "packages": []
        }
    ],
    "total": 2,
    "totalInactive": 0,
    "page": 1,
    "perPage": 20
}
```
</details>

criamico commented:
> I'm having trouble getting past the "Confirm incoming data" step in Cloud.
> 
> I'm enrolling an agent in a basic Agent Policy that just includes the `system` integration, and I'm running the agent on my Macbook via docker, e.g.
> 
> ```
> docker run --rm --add-host elasticsearch.local:192.168.65.2 \
>   -e FLEET_URL=https://6e3f3c836ed0d22d3db21117a0894f2f.fleet.us-west2.gcp.elastic-cloud.com:443\
>   -e FLEET_ENROLL=true \
>   -e FLEET_ENROLLMENT_TOKEN=MThjV0JJQUJFTENteXpEd0xBTDI6ZUZSRHpvVmlTUGVPWVFHWkhHTFZ3Zw== \
>   docker.elastic.co/beats/elastic-agent:8.2.0-SNAPSHOT
> ```
> 
> ![image](https://user-images.githubusercontent.com/6766512/162747506-7752677e-ed10-4605-aecf-05e4c9b84e3e.png)
> 
> My agent starts up as expected, and Fleet reports that it's enrolled, but the "Confirm incoming data" spinner never goes away. I let it sit for 15ish minutes with no change.
> 
> Here's what I think is the relevant polling request from my network tab:
> 
> `/api/fleet/agents?page=1&perPage=20&showInactive=false&showUpgradeable=false`
> I do see system/metrics data in my `Metrics System Overview` dashboard for the agent, and logs in its `Logs` tab, so I'm not sure what's missing to get past the loader.

Ok, so there's still some problem there. The polling request is similar to this one`/api/fleet/agent_status/data?agentsIds=cf852616-61d9-4318-bf9f-fcfb3d72042d` and should happen every 5s until some data is retrieved. Here's a short video of the functionality on my local:

https://user-images.githubusercontent.com/16084106/162750572-6fb38f9e-5264-42f1-953e-04436b82ed40.mov

The other thing that I had to do to see the data is to check that the fleet server host in the settings has the correct IP (http://DOCKERHOST_IP:8220) and I also pass it to the docker command (in addition to your command):

`-e FLEET_URL=http://$DOCKERHOST_IP:8220 \`

criamico commented:
@juliaElastic @kpollich I added a check that should solve the problem with the fail retries in the incoming data component. I'm filtering out agents that have`{data: false}` in the response to `/api/fleet/agent_status/data?agentsIds={agent_ids}` query and then I'm setting `incomingData` to the filtered array; this way it should keep polling until it gets valid data.

juliaElastic commented:
> @juliaElastic @kpollich I added a check that should solve the problem with the fail retries in the incoming data component. I'm filtering out agents that have`{data: false}` in the response to `/api/fleet/agent_status/data?agentsIds={agent_ids}` query and then I'm setting `incomingData` to the filtered array; this way it should keep polling until it gets valid data.

I tested and the polling works for me now, eventually I see the data confirmed on the flyout.

juliaElastic commented on x-pack/plugins/fleet/public/components/agent_enrollment_flyout/use_get_agent_incoming_data.tsx:
```diff
@@ -92,8 +92,16 @@ export const usePollingIncomingData = (agentsIds: string[]) => {
-          setIsLoading(false);
+          // filter out agents that have `data = false` and keep polling
+          const filtered = data?.items.filter((item) => {
+            const key = Object.keys(item)[0];
```
in case of multiple `agentIds` being queried, do they come in separate objects in `items` array? so we are not missing anything by taking the first key only, right?

criamico replied:
No there's no problem there, the schema is something like
`[
    {
       'agentid1': {
           data: true
    },
   {
       'agentid2': {
           data: false
    },
    ...
]`

juliaElastic reviewed (approved):
LGTM, thanks for fixing the review comments

criamico commented:
> LGTM, thanks for fixing the review comments

Thanks a lot for you help! It's better to catch those issues early :)
